drivers: gpio: Add Ambiq GPIO driver

aaronyegx opened the pull request:
This PR adds the Ambiq GPIO driver, as well as instances in the SoC dtsi file, enabling it on the apollo4p_evb board.
It also adds support for LEDs and buttons.
Most of the driver code are picked from another long-delayed PR https://github.com/zephyrproject-rtos/zephyr/pull/61157, which includes the general Ambiq GPIO driver and Ambiq connecter/header for shield boards. This PR only includes the general GPIO driver in order to move forward the GPIO driver implementation to completion.
We updated the Ambiq GPIO instance by using GPIO maps in devicetree to include all GPIOs, which was suggested in the https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1679490658 and https://github.com/zephyrproject-rtos/zephyr/pull/61157#discussion_r1393267928, also updated the GPIO driver accordingly.

Tested the following samples on apollo4p_evb:
- samples/basic/blinky.
- samples/basic/threads.
- samples/basic/button.

aaronyegx commented:
Hi @henrikbrixandersen , thanks for your review and comment on the PR https://github.com/zephyrproject-rtos/zephyr/pull/61157 which includes the general Ambiq GPIO driver and Ambiq connecter/header for shield boards. I picked the general GPIO driver codes from there and addressed your comments https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1679490658 and https://github.com/zephyrproject-rtos/zephyr/pull/61157#discussion_r1393267928 put on that PR. Can you help take another look at this?
We are in a hurry for this feature since we are close to the deadline of some projects and trying our best to align the codes with the Zephyr specification. Thanks for your understanding.

henrikbrixandersen commented on drivers/gpio/gpio_ambiq.c:
```diff
@@ -0,0 +1,359 @@
+		pincfg.GP.cfg_b.eIntDir = AM_HAL_GPIO_PIN_INTDIR_NONE;
+		ret = am_hal_gpio_pinconfig(gpio_pin, pincfg);
+
+		irq_disable(dev_cfg->irq_num);
```
Won't this disable IRQs for _all_ pins?

aaronyegx replied:
Thanks for catching this. Yes, we cannot call irq_disable() for the GPIO pin interrupt disablement here since every 32 pins share the same irq number.

henrikbrixandersen commented on dts/bindings/gpio/ambiq,gpio.yaml:
```diff
@@ -0,0 +1,19 @@
+include: [gpio-controller.yaml, base.yaml]
+
+properties:
+  reg:
+    required: true
```
Please add a description of how the `reg` property is used for the child devices and its relationship with the parent `reg` property.

gmarull commented on dts/arm/ambiq/ambiq_apollo4p.dtsi:
```diff
@@ -206,6 +207,61 @@
+			#size-cells = <0>;
+
+			gpio: gpio@40010000 {
+				compatible = "simple-bus";
```
?

aaronyegx replied:
Thanks for your catching this. Did cleanup.

msierszulski committed:
drivers: gpio: add Ambiq GPIO driver

This commit adds GPIO driver for Apollo4 SoCs.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
boards: arm: apollo4p_evb: Enable buttons

This commit adds buttons instances and aliases for them
on apollo4p_evb boards.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
dts: arm: ambiq: Add GPIO instances to SoC

This commit instantiates the GPIO peripherals.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
boards: arm: apollo4p_evb: Enable LEDs

This commit adds leds instances and aliases for them
on apollo4p_evb board.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

aaronyegx committed:
drivers: gpio: ambiq: Fix the incorrect IRQ disablement.

For the Ambiq Apollo4x soc, every 32 pins share the same IRQ
number. irq_disable() should not be called for the pin interrupt
disablement, otherwise the interrupt of pins in the same GPIO
group will be disabled as well.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx commented:
Hi @henrikbrixandersen @gmarull , since Ambiq Apollo4x soc hardware only defines one GPIO port with 128 pins, I created one "ambiq,gpio-base" binding for Apollo4x soc devicetree to define one GPIO "parent" node. Then we can create multiple child nodes which are compatible with "ambiq,gpio" under the "ambiq,gpio-base" parent node to solve the bitmask limitation of GPIO driver API. I provided the description and example in the binding file. Could you help check if it is acceptable? Thank you.

henrikbrixandersen commented on dts/bindings/gpio/ambiq,gpio-base.yaml:
```diff
@@ -0,0 +1,84 @@
+    = 0x400100D0 and the real GPIO pin number of this pin in soc is (20 + 32)
+    = 52.
+
+compatible: "ambiq,gpio-base"
```
We typically reserve the `*-base` bindings for files, that are to be included in other bindings, but do not include their own `compatible` string.

The scheme is perfectly fine, though. However, I'd suggest naming the top-level compatible `ambiq,gpio` and the child compatibles `ambiq,gpio-bank` (or any similar term that may fit in better with the datasheet terminology). This avoid the somewhat confusing use of `*-base` in a binding compatible.

aaronyegx replied:
@henrikbrixandersen thanks for your review. I updated the GPIO binding files ("ambiq,gpio" for the parent node and "ambiq,gpio-bank" for the child node).

aaronyegx committed:
dts: binding: gpio: Add Ambiq gpio and gpio-bank bindings.

This Ambiq gpio binding provides the GPIO pin mapping for GPIO child
nodes tosolve the limitation of the maximum 32 pins handling in GPIO
driver API.
The gpio-bank nodes can be created under the gpio parent node.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx committed:
dts: arm: ambiq: Update the GPIO instances

Use the "ambiq,gpio" binding to combine the "ambiq,gpio-bank"
child nodes for Apollo4 Plus soc.
Also update the GPIO driver accordingly.

Signed-off-by: Aaron Ye <aye@ambiq.com>

henrikbrixandersen reviewed (approved):
Looks good.

fkokosinski reviewed (approved):
LGTM
